Redo developer install steps for happy path

mrjones-plip opened the pull request:
per #700, add a new doc that simplifies [the current developer](https://github.com/medic/cht-core/blob/575bc95a8ccf82b211a13fec9d3228f71169405f/DEVELOPMENT.md) hosting doc to easy, copy and paste-able steps.

mrjones-plip commented:
Thanks for [the feedback](https://github.com/medic/cht-docs/issues/700#issuecomment-1139631863) in the parent ticket Gareth and Craig! I opted to include all the content from the current [developer doc](https://github.com/medic/cht-core/blob/575bc95a8ccf82b211a13fec9d3228f71169405f/DEVELOPMENT.md), but just front load the happy path, and literally call it the happy path.

If this is too verbose, happy to trim it down and we can add it back later.  If "the happy path" is too on the nose for titles, open to suggestions to change it!

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+First, update your current packages:
+
+```shell
+sudo apt update&&sudo apt -y dist-upgrade
````
Is the "dist-" part necessary? I'd have just expected a `sudo apt upgrade`

mrjones-plip replied:
Based on my reading [here](https://linoxide.com/difference-between-apt-get-upgrade-and-dist-upgrade/) and [here](https://askubuntu.com/a/226213), it is now a best practice for me for all but production instances (of which I run none b/c personal servers are not "production" ;) . I think it fits here in dev land, so resolving, but happy to get some push back!

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+```shell
+curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
+. ~/.bashrc
+nvm install v12
````
We can save money on that unnecessary 'v' character ;)

```suggestion
nvm install 12
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+Then install `nvm`, add it to your path and install NodeJS 12:
+
+```shell
+curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | bash
+. ~/.bashrc
````
This will work for both zsh and bash.
```suggestion
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | $0
. ~/.$0rc
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+Now let's ensure NodeJS 12 and npm 6 were installed. This should output version 12.x.x for NodeJS and 6.x.x for `npm`:
+
+```shell
+node -v&&npm -v
````
Spaces for readability
 ```suggestion
node -v && npm -v
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+sudo apt -y install xsltproc curl uidmap jq
+```
+
+Install Docker with this sweet one liner (note it's directly from docker.com):
````
The other text is an unnecessary detail
```suggestion
Install Docker:
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+
+```shell
+dockerd-rootless-setuptool.sh install
+echo "export PATH=/usr/bin:$PATH" >> ~/.bashrc
+echo "export DOCKER_HOST=unix:///run/user/1000/docker.sock" >> ~/.bashrc
+. ~/.bashrc
````
Please consider us zsh people! :)
```suggestion
echo "export PATH=/usr/bin:$PATH" >> ~/.$0rc
echo "export DOCKER_HOST=unix:///run/user/1000/docker.sock" >> ~/.$0rc
. ~/.$0rc
```

mrjones-plip replied:
Just FYI:

```shell
➜  cht-docs git:(700-developer-install-happy-path) echo $0
zsh
```

(technically oh-my-z-sh)

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+
+### CHT Core cloning and setup
+
+Clone the main CHT Core repo from GitHub and initialize it with an `npm` command. Note this last command may take many minutes. Be patient!
+
+```shell
+git clone https://github.com/medic/cht-core ~/cht-core
+cd ~/cht-core
+npm ci  
````
Please split out the `npm ci` call and instructions for it. It's different enough from the git clone. Then you can give a bit more detail on the `npm ci` with something like "install dependencies and perform other setup tasks".

mrjones-plip replied:
yup yup - good call! done in 3f4f766

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,235 @@
+
+### Developing
+
+Now you have everything installed and can begin development! You'll need three separate terminals when doing development. In the first terminal, run:
```
>You'll need three separate terminals
I always appreciate things like this. It tells me what's about to happen and then when I'm doing it I know the context of what I'm doing. I know this is a really small case, but I still really like it :)

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+Then in a 2nd terminal run:
+
+```shell
+cd ~/cht-core&&grunt dev-api
````
Spaces
```suggestion
cd ~/cht-core && grunt dev-api
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+
+How to permanently define environment variables depends on your OS and shell (e.g. for bash you can put them `~/.bashrc`). You can temporarily define them with `export`:
+
+```sh
````
```suggestion
```shell
```

craig-landry commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,235 @@
+
+## nginx-local-ip
+
+[`nginx-local-ip`](https://github.com/medic/nginx-local-ip) is a local proxy that keeps all traffic local, and runs without latency or throttling. If sharing your local CHT instance is not required, it is the preferred method to add a valid SSL certificate (rather than `ngrok` or `pagekite`).
```
Small wording changes
```suggestion
[`nginx-local-ip`](https://github.com/medic/nginx-local-ip) is a local proxy that keeps all traffic local, and runs without latency or throttling. If sharing your local CHT instance is not required, it is the recommended method to add a valid SSL certificate (rather than `ngrok` or similar).
```

mrjones-plip commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,235 @@
+
+```shell
+npm install -g grunt-cli
+sudo apt -y install xsltproc curl uidmap jq
````
Agreed! fixed in 6c9961e

mrjones-plip commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,235 @@
+* [npm 6.x.x](https://npmjs.com/)
+* [grunt cli](https://gruntjs.com/using-the-cli)
+* [xsltproc](http://www.sagehill.net/docbookxsl/InstallingAProcessor.html) 
+* [python 2.7](https://www.python.org/downloads/)
```
Gotcha. I dug up [this ticket](https://github.com/medic/cht-core/issues/5608) which says maybe it won't work? I don't want to solve this now given how far reaching it might be.  I say we leave it as is and do nothing or leave it as is and file a "update requirements to be python 3" ticket.

Thoughts?

mrjones-plip replied:
Meh - doing [the latter](https://github.com/medic/cht-core/issues/7636).

mrjones-plip commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,235 @@
+1. Create a [pagekite account](https://pagekite.net/signup/), download and install the python script.
+1. Start pagekite (be sure to replace `YOUR-PAGEKIT-NAME` with the URL you signed up for) to connect to:
+  * The CHT API running via `grunt` or `horti`, execute `python2 pagekite.py 5988 YOUR-PAGEKIT-NAME.pagekite.me`
+  * The CHT API running via `docker`, execute `python2 pagekite.py 443 YOUR-PAGEKIT-NAME.pagekite.me`
```
[nope](https://pagekite.net/downloads#s3linux)! fixed in c003704

mrjones-plip commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,235 @@
+- `COUCH_URL`: the full authenticated url to the `medic` DB. Locally this would be  `http://myadminuser:myadminpass@localhost:5984/medic`
+- `COUCH_NODE_NAME`: the name of your CouchDB's node. The Docker image default is `nonode@nohost`. Other installations may use `couchdb@127.0.0.1`. You can find out by querying [CouchDB's membership API](https://docs.couchdb.org/en/stable/api/server/common.html#membership)
+- (optionally) `API_PORT`: the port API will run on. If not defined we use `5988`
+- (optionally) `CHROME_BIN`: only required if `grunt unit` or `grunt e2e` complain that they can't find Chrome.
```
fixed in 8752c44

garethbowen commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,240 @@
+Then install `nvm`, add it to your path and install NodeJS 12:
+
+```shell
+curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.34.0/install.sh | $0
````
`0.34.0` is not the latest and the docs now recommend using `0.39.1`. But I wonder if we should just say `master` so we don't have to update this every time? Feels wrong though.

mrjones-plip replied:
Hmm - `curl` + `jq` = ftw? see 882adb2

garethbowen commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,240 @@
+sudo apt -y install xsltproc curl uidmap jq
+```
+
+Then install `nvm`, add it to your path and install NodeJS 12:
````
How much do we want to explain _why_ we're doing this? Like, what even is nvm?

garethbowen commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,240 @@
+
+```shell
+docker run hello-world
+```
````
I love these verification steps. Fail early!

garethbowen commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,240 @@
+echo "export COUCH_NODE_NAME=nonode@nohost">> ~/.$0rc
+echo "export COUCH_URL=http://medic:password@localhost:5984/medic">> ~/.$0rc
+. ~/.$0rc
+echo $COUCH_NODE_NAME && echo $COUCH_URL
```
Maybe split this out as a separate block so we can say "to make sure it's working, run this:"

mrjones-plip replied:
yup yup! done in 39f1bbb

garethbowen commented on content/en/apps/guides/hosting/core-developer.md:
```diff
@@ -0,0 +1,240 @@
+
+When you're done with development you can `ctrl + c` in the three terminals and stop the CouchDB container with `docker stop medic-couchdb`.  When you want to resume development later, run `docker start medic-couchdb` and re-run the three terminal commands.
+
+## Other Paths
```
Maybe call this "Troubleshooting"? If you've forgotten (like I had) that the first heading was "Happy Path" this heading makes no sense.

mrjones-plip replied:
renamed AND linked - let's help find the right path! see 5a7ae9a

garethbowen reviewed (changes requested):
Great work. I didn't read every line, but what I read I liked!  :D

garethbowen reviewed (approved):
LGTM. Probably not perfect, but much better than what we have now, and we can iterate from here. Thanks for the lift!

mrjones-plip commented:
Rad - thanks for the review! That was a good catch on the `nvm` installer - and TIL about that GH API for getting the latest version.

Craig - I'll wait to hear back from you before merging b/c you had so much helpful feedback - I wanna be sure it's satisfactorily resolved.

mrjones-plip commented:
Note to self, tend to the links to the old developer guide too as found with `grep -r 'DEVELOPMENT.md' content/en`:

* **content/en/core/guides/using-windows.md**: From this point, follow the setup guide from `Enabling a secure CouchDB` section in [Development Guide](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md).
content/en/core/guides/android/development-setup.md:Refer to the [CHT Core Developer Guide](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md#testing-locally-with-devices).
* **content/en/apps/guides/debugging/secure-sharing-of-developer-instance.md:**  When using a local [development environment](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md), you may want to share your work with other collaborators. You also may need to access the environment from your [mobile device](https://github.com/medic/cht-android) which requires an SSL certificate (the "s" in "https"). By using a publicly accessible web server, you can receive the secure https requests and forward them back to your CHT instance which doesn't have https set up:
* **content/en/apps/guides/debugging/secure-sharing-of-developer-instance.md:**  You have a local [dev instance](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md) set up of cht-core 
* **content/en/apps/guides/debugging/replicating-production-data-locally.md:**  If you don't already have a local dev env, follow the instructions on the [development setup instructions](https://github.com/medic/cht-core/blob/master/DEVELOPMENT.md).

mrjones-plip commented on content/en/apps/guides/hosting/core-developer.md:
````diff
@@ -0,0 +1,245 @@
+
+```shell
+sudo apt update && sudo apt -y dist-upgrade
+sudo apt -y install xsltproc curl uidmap jq
````
ok - rad - I'll commit this and then do another run through to make sure it all gels together.

craig-landry replied:
> ok - rad - I'll commit this and then do another run through to make sure it all gels together.

Thanks. I'll look for your confirmation that it all actually works and then I'm good to approve.

mrjones-plip commented:
Whew - testing got a LOT better when I started [using Mokhtar's branch](https://github.com/medic/cht-core/tree/fix-npm-ci-hanging), taking testing down to 5 min per `grunt` and `npm ci`  to about 1 min. Before:

```
Running "exec:npm-ci-modules" (exec) task                                                                                                                                                                          
[webapp]                                                                                                                                                                                                           
added 66 packages in 6.832s                                                                                                                                                                                        
[api]                                                                                                                                                                                                              
added 432 packages in 395.736s                                                                                                                                                                                     
[sentinel]                                                                                               
added 127 packages in 1.456s                                                                                                                                                                                       
[admin]   
```

After:

```
Running "exec:npm-ci-modules" (exec) task
[webapp]
added 66 packages in 5.777s
[api]
added 432 packages in 3.579s
[sentinel]
added 127 packages in 1.524s
[admin]
```

Craig - I added `apt install python2`  and everything went well on my happy path.  How to do I test if python 2 vs 3 is needed?

mrjones-plip commented:
Oh no - sorry Gareth - Didn't mean to re-re-request review!

craig-landry commented:
> How to do I test if python 2 vs 3 is needed?

I don't know all the cases. For now that doesn't need to block this though. We can make that switch later, in a separate PR.
